monix-eval/shared/src/main/scala/monix/eval/internal/TaskCreate.scala

```diff
@@ -226,10 +226,11 @@ private[eval] object TaskCreate {
       if (shouldPop) ctx.connection.pop()
       // Optimization — if the callback was called on the same thread
       // where it was created, then we are not going to fork
+      // This is not safe to do when localContextPropagation enabled
```

Comment:
Why is it not safe?

Reply:
Because `this` extends a `TrampolinedRunnable`, and `TracingScheduler` extends `BatchingScheduler`, doing `execute` bypasses the `TracingRunnable` creation and context isolation that it has.

Without that isolation, the continuation of that async/asyncF/etc. node therefore will have the context from whoever called the `cb`, not from where the `async` was actually created.

That only happened in case `cb` was called synchronously, but it does happen deep in ConcurrentChannel impl in certain concurrent usage patterns with fast producer and slow-ish consumer.